apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
+    return navigationTransition.defer.promise.then(
+      () => startNavigation(opts)
+    );
+  }
+
+  if (!navigationTransition) {
+    var defer = Utils.Promise.defer();
+
+    navigationTransition = {
+      stepsRun: [],
+      defer
+    };
+  }
+
+  navigationTransition.partial = partial;
+
+  if (args && ('meta' in args)) {
+    throw new Error('Arguments cannot have a `meta` property.');
+  }
+
+  navigationTransition.transitionArgs = {
+    meta: {
+      next: newPanel ? {
+        panel: newPanel.name,
+        args
+      } : null,
+      prev: currentState ? {
+        panel: currentState.panel.name,
+        args: currentState.args
+      } : null
+    }
+  };
+
+  Object.assign(navigationTransition.transitionArgs, args);
+
+  navigationTransition.states = {
+    from: currentState,
+    to: {
+      panel: newPanel,
+      args
+    }
+  };
+
+  return Promise.resolve();
+}
+
+/**
+ * Sets the currentState variable to its new value.
+ */
+function setFutureState() {
+  currentState = navigationTransition.states.to;
+}
+
+/**
+ * Ends the navigation so that new navigations can take place.
+ * This cleans up the navigation states.
+ */
+function endNavigation() {
+  debug('endNavigation()');
+  navigationTransition && navigationTransition.defer.resolve();
+  navigationTransition = null;
+}
 
-// This function is used as a rejection handler, to swallow errors and log them
-function catchError(e) {
+/**
+ * This function is used to catch errors at the end of navigation promise
+ * chains. We're ending the navigation if this happens, so that we can start a
+ * new one afterwards.
+ *
+ * @param {Exception} [e] The exception to display, if any.
+ */
+function onNavigationError(e) {
+  console.error('Navigation error', e);
+  endNavigation();
+}
+
+/**
+ * This function is used as a rejection handler, to swallow errors and log them.
+ *
+ * @param {Exception} [e] The exception to display, if any.
+ */
+function catchStepError(e) {
   if (e) {
     console.error(
       'Navigation.toPanel: got an exception: (%s) %s',
       e, e.stack && e.stack.replace(/\n/g, '|')
     );
   } else {
-    console.log('Navigation.toPanel: got a rejected promise');
+    debug('Navigation.toPanel: got a rejected promise');
   }
 }
 
-var currentPanel;
-var queuedPanel;
+/**
+ * Executes the lifecycle function for `stepName` on the right object.
+ *
+ * @param {String} stepName The step to execute: one of 'beforeLeave',
+ * 'beforeEnter', 'afterLeave', 'afterEnter'.
+ * @returns {Promise} Resolved when the step returns a value or a resolved
+ * promise, rejected when the step either throws or returns a rejected promise.
+ * Or a resolved promise if no step was executed (because it was already
+ * executed or there was no function to execute).
+ */
+function executeNavigationStep(stepName) {
+  debug(`executeNavigationStep(${stepName})`);
+  if (!navigationTransition) {
+    throw new Error(
```

Comment:
Again, not a big deal right now, I'd say it's a bit better for consistency, I just see that below we return ```Promise.resolve()``` instead of just ```return;```.